🐛 fix: support `PROXY_URL` in docker with proxychains-ng

hezhijie0327 opened the pull request:
#### 💻 变更类型 | Change Type

<!-- For change type, change [ ] to [x]. -->

- [ ] ✨ feat
- [ ] 🐛 fix
- [ ] ♻️ refactor
- [ ] 💄 style
- [ ] 🔨 chore
- [ ] ⚡️ perf
- [x] 👷 build
- [x] 📝 docs

#### 🔀 变更说明 | Description of Change
1. 以 Xray 的 Socks5 为例，配置 `PROXY_URL="socks5://127.0.0.1:7891"`
2. 增加 Alpine Package Mirror，加速构建 [TODO: 镜像源当前为 USTC，是否有必要改为其他？]
3. 重命名 `USE_NPM_CN_MIRROR` 为 `USE_CN_MIRROR`
<!-- Thank you for your Pull Request. Please provide a description above. -->

- close #2004 
- close #1936 
- close #1990 
- close #1613 
- close #1303 
- close #985 

#### 📝 补充信息 | Additional Information
![image](https://github.com/user-attachments/assets/14888394-5e82-46fc-9059-c0deb64855d5)
![image](https://github.com/user-attachments/assets/6649b66b-9d64-4a00-801c-2f77fe7b7c20)

replace: https://github.com/lobehub/lobe-chat/pull/2382
<!-- Add any other context about the Pull Request here. -->

hezhijie0327 committed:
✨: feat: support PROXY_URL in docker with proxychains-ng

hezhijie0327 committed:
🔨 chore: let proxychains config more readable

hezhijie0327 committed:
🔨 chore: add package update

hezhijie0327 committed:
⚡️ perf: add CN mirror for Alpine

hezhijie0327 committed:
🔨 chore: add more comments

hezhijie0327 committed:
🔨 chore: change USE_NPM_CN_MIRROR to USE_CN_MIRROR

hezhijie0327 committed:
🐛 fix: fix cn mirror for alpine during building

hezhijie0327 committed:
🐛 fix: fix permission issue

hezhijie0327 committed:
🐛 fix: fix app path error

hezhijie0327 commented:
TODO:
1. 加入 `nslookup localhost | grep 'Address:' | tail -n 1 | awk 'print $2'` 支持域名格式服务器解析
2. A 及 AAAA 记录，`-q=A` 或 `-q=AAAA`
3. 当解析结果为空时，返回输入值

hezhijie0327 committed:
✨ feat: support proxy server with domain

hezhijie0327 committed:
🔨 chore: add more comments & cleanup code

hezhijie0327 committed:
🔨 chore: add more comments

hezhijie0327 commented:
@arvinxx 代理功能都实现了，可以试试

构建: 
`docker build -t lobe-chat:local --build-arg USE_CN_MIRROR=true --no-cache .`

代理：
`docker run -it --rm -p 3210:3210 -e PROXY_URL="socks5://127.0.0.1:7891" lobe-chat:local`

代理带域名：
`docker run -it --rm -p 3210:3210 -e PROXY_URL="socks5://localhost:7891" lobe-chat:local`

hezhijie0327 committed:
⚡️ perf: resolve only when host is domain

hezhijie0327 committed:
🔨 chore: move IP_REGEX to head

arvinxx commented:
@hezhijie0327 必须要带 `USE_CN_MIRROR` 才能打出 PROXY_URL 的版本吗？

arvinxx commented:
哦  看了下，直接配 PROXY_URL 就行

hezhijie0327 commented:
嗯 `USE_CN_MIRROR` 这个仅在镜像构建中用到，主要是 Alpine 也需要镜像干脆就统一一个变量了

hezhijie0327 committed:
🔨 chore: replace `! -z` to `-n`

hezhijie0327 committed:
🔨 chore: update comments

yexingok commented:
感谢作者，请问这个功能啥时候回合并到主线？
sock5代理恰好是一个需要的功能

hezhijie0327 committed:
✨ feat: add `DEFAULT_AGENT_CONFIG` and `SYSTEM_AGENT` to env, allow set default model

arvinxx commented:
@hezhijie0327 我测下来似乎不行？我用的代理是 ClashX



<img width="1343" alt="image" src="https://github.com/user-attachments/assets/bc0c688f-6b0c-4472-a7ba-38ba02a47171">

hezhijie0327 commented:
@arvinxx 是 `Docker Desktop` 嘛?

如果是的话，建议用当前计算机的实际 IP 地址（192.168.0.x）或容器的 Host IP，并开启局域网访问（可能是叫这个名字，类似这种功能，需要将服务绑定在 0.0.0.0/0 上 ） 

`Docker Desktop` 在 `Windows `和 `macOS `上走的是虚拟机方案，如果是 `localhost` 的话是走到 `lobe-chat` 自身容器的 `localhost`

hezhijie0327 commented:
尝试用 `host.docker.internal` 替代 `localhost`

`docker run -it --rm -p 3210:3210 -e PROXY_URL="socks5://host.docker.internal:7890" lobe-chat:local`

![image](https://github.com/user-attachments/assets/f6e80c44-efe0-4bb5-af49-750facc2b4d9)

arvinxx commented:
测下来没问题了，https://github.com/hezhijie0327/lobe-chat/pull/43 补了下文档

arvinxx committed:
Add documentation for `PROXY_URL` environment variable

Related to #2004

Add documentation for `PROXY_URL` environment variable in both English and Chinese versions of the environment variables guide.

* **English Documentation**
  - Add `PROXY_URL` section to specify the proxy URL for connecting to external services.
  - Include example values and a callout for Docker Desktop users.

* **Chinese Documentation**
  - Add `PROXY_URL` section to specify the proxy URL for connecting to external services.
  - Include example values and a callout for Docker Desktop users.

---

For more details, open the [Copilot Workspace session](https://copilot-workspace.githubnext.com/lobehub/lobe-chat/pull/3362?shareId=XXXX-XXXX-XXXX-XXXX).

hezhijie0327 commented:
好的，merge 了

zhuozhiyongde commented:
我使用 `PROXY_URL=http://172.17.0.1:7890` 好像不行，以下是日志：

```shell
docker logs lobe-database -f
🌐 DNS Server: [ '127.0.0.11' ]
-------------------------------------
[Database] Start to migration...
✅ database migration pass.
-------------------------------------
✅ ProxyChains: All outgoing traffic routed via http://172.17.0.1:7890.
-------------------------------------
   ▲ Next.js 15.3.5
   - Local:        http://localhost:3210
   - Network:      http://0.0.0.0:3210

 ✓ Starting...
 ✓ Ready in 202ms
Warning: Cannot load "@napi-rs/canvas" package: "Error: Cannot find module '@napi-rs/canvas'
Require stack:
- /app/node_modules/.pnpm/pdfjs-dist@4.10.38/node_modules/pdfjs-dist/legacy/build/pdf.mjs".
Warning: Cannot polyfill `DOMMatrix`, rendering may be broken.
Warning: Cannot polyfill `ImageData`, rendering may be broken.
Warning: Cannot polyfill `Path2D`, rendering may be broken.
Error in tRPC handler (edge) on path: market.getPluginIndex
TRPCError: No procedure found on path "market.getPluginIndex"
    at <unknown> (.next/server/app/(backend)/trpc/edge/[trpc]/route.js:1:20594)
    at Array.map (<anonymous>)
    at eF (.next/server/app/(backend)/trpc/edge/[trpc]/route.js:1:20521)
    at async eV (.next/server/app/(backend)/trpc/edge/[trpc]/route.js:5:3967)
    at async eh.do (.next/server/edge-chunks/509.js:5:11300)
    at async eh.handle (.next/server/edge-chunks/509.js:5:13369)
    at async eE.handler (.next/server/edge-chunks/509.js:17:31222)
    at async M (.next/server/edge-chunks/509.js:17:4974) {
  cause: undefined,
  code: 'NOT_FOUND'
}
Provider runtime map found for provider: google
Error: exception TypeError: fetch failed sending request
    at <unknown> (.next/server/edge-chunks/406.js:6:159621)
    at async tz.generateContentStream (.next/server/edge-chunks/406.js:6:197363)
    at async ii.chat (.next/server/edge-chunks/938.js:15:15907)
    at async (.next/server/app/(backend)/webapi/chat/google/route.js:1:10174)
    at async eh.do (.next/server/edge-chunks/509.js:5:11300)
    at async eh.handle (.next/server/edge-chunks/509.js:5:13369)
    at async eE.handler (.next/server/edge-chunks/509.js:17:31222)
    at async M (.next/server/edge-chunks/509.js:17:4974) {
  
}
Route: [google] ProviderBizError: { message: 'exception TypeError: fetch failed sending request' }
```

```shell
sudo lsof -i:7890  
COMMAND    PID   USER   FD   TYPE   DEVICE SIZE/OFF NODE NAME
clash   255877 ubuntu    7u  IPv6 78955401      0t0  TCP *:7890 (LISTEN)
```

zhuozhiyongde commented:
换成 socks 端口好像也不行：

```shell
docker logs lobe-database -f
🌐 DNS Server: [ '127.0.0.11' ]
-------------------------------------
[Database] Start to migration...
✅ database migration pass.
-------------------------------------
✅ ProxyChains: All outgoing traffic routed via socks5://172.17.0.1:7891.
-------------------------------------
   ▲ Next.js 15.3.5
   - Local:        http://localhost:3210
   - Network:      http://0.0.0.0:3210

 ✓ Starting...
 ✓ Ready in 197ms
Warning: Cannot load "@napi-rs/canvas" package: "Error: Cannot find module '@napi-rs/canvas'
Require stack:
- /app/node_modules/.pnpm/pdfjs-dist@4.10.38/node_modules/pdfjs-dist/legacy/build/pdf.mjs".
Warning: Cannot polyfill `DOMMatrix`, rendering may be broken.
Warning: Cannot polyfill `ImageData`, rendering may be broken.
Warning: Cannot polyfill `Path2D`, rendering may be broken.
Error in tRPC handler (edge) on path: market.getAgentIndex
TRPCError: No procedure found on path "market.getAgentIndex"
    at <unknown> (.next/server/app/(backend)/trpc/edge/[trpc]/route.js:1:20594)
    at Array.map (<anonymous>)
    at eF (.next/server/app/(backend)/trpc/edge/[trpc]/route.js:1:20521)
    at async eV (.next/server/app/(backend)/trpc/edge/[trpc]/route.js:5:3967)
    at async eh.do (.next/server/edge-chunks/509.js:5:11300)
    at async eh.handle (.next/server/edge-chunks/509.js:5:13369)
    at async eE.handler (.next/server/edge-chunks/509.js:17:31222)
    at async M (.next/server/edge-chunks/509.js:17:4974) {
  cause: undefined,
  code: 'NOT_FOUND'
}
Provider runtime map found for provider: google
Error in tRPC handler (edge) on path: config.getGlobalConfig
TRPCError: No procedure found on path "config.getGlobalConfig"
    at <unknown> (.next/server/app/(backend)/trpc/edge/[trpc]/route.js:1:20594)
    at Array.map (<anonymous>)
    at eF (.next/server/app/(backend)/trpc/edge/[trpc]/route.js:1:20521)
    at async eV (.next/server/app/(backend)/trpc/edge/[trpc]/route.js:5:3967)
    at async eh.do (.next/server/edge-chunks/509.js:5:11300)
    at async eh.handle (.next/server/edge-chunks/509.js:5:13369)
    at async eE.handler (.next/server/edge-chunks/509.js:17:31222)
    at async M (.next/server/edge-chunks/509.js:17:4974) {
  cause: undefined,
  code: 'NOT_FOUND'
}
Error: exception TypeError: fetch failed sending request
    at <unknown> (.next/server/edge-chunks/406.js:6:159621)
    at async tz.generateContentStream (.next/server/edge-chunks/406.js:6:197363)
    at async ii.chat (.next/server/edge-chunks/938.js:15:15907)
    at async (.next/server/app/(backend)/webapi/chat/google/route.js:1:10174)
    at async eh.do (.next/server/edge-chunks/509.js:5:11300)
    at async eh.handle (.next/server/edge-chunks/509.js:5:13369)
    at async eE.handler (.next/server/edge-chunks/509.js:17:31222)
    at async M (.next/server/edge-chunks/509.js:17:4974) {
  
}
Route: [google] ProviderBizError: { message: 'exception TypeError: fetch failed sending request' }
Provider runtime map found for provider: google
Error: exception TypeError: fetch failed sending request
    at <unknown> (.next/server/edge-chunks/406.js:6:159621)
    at async tz.generateContentStream (.next/server/edge-chunks/406.js:6:197363)
    at async ii.chat (.next/server/edge-chunks/938.js:15:15907)
    at async (.next/server/app/(backend)/webapi/chat/google/route.js:1:10174)
    at async eh.do (.next/server/edge-chunks/509.js:5:11300)
    at async eh.handle (.next/server/edge-chunks/509.js:5:13369)
    at async eE.handler (.next/server/edge-chunks/509.js:17:31222)
Route: [google] ProviderBizError: { message: 'exception TypeError: fetch failed sending request' }
    at async M (.next/server/edge-chunks/509.js:17:4974) {
  
}
```

```shell
sudo lsof -i:7891
COMMAND    PID   USER   FD   TYPE   DEVICE SIZE/OFF NODE NAME
clash   255877 ubuntu    8u  IPv6 78955402      0t0  TCP *:7891 (LISTEN)
clash   255877 ubuntu    9u  IPv6 78955403      0t0  UDP *:7891 
```

Clash config：

<img width="818" height="526" alt="image" src="https://github.com/user-attachments/assets/a90dd88c-fefa-4806-ae21-8e3f8df77181" />

zhuozhiyongde commented:
参考 https://stackoverflow.com/questions/74796374/mediasoup-error-when-running-observertc-example-getaddrinfo-enotfound-host-doc

host.docker.internal 可能在 Linux 系统上无法正确解析，而是应当使用 172.18.0.1 或添加 extra-host

zhuozhiyongde commented:
更新：添加 extra-host 之后，解析看起来正常了，没有继续报 DNS 错误，但是流量无法正确被代理到宿主机 clash 进程，clash 日志中没有显示连接。
